### DISCUSSION TITLE
bug: using nextjs app and pages dir breaks typescript language server

### DISCUSSION THREAD

**Discussion by matheins** [trpc/trpc#5193] [Q&A]
### Provide environment information

```
System:
    OS: macOS 13.3
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 426.82 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 19.8.1 - ~/.nvm/versions/node/v19.8.1/bin/node
    npm: 9.5.1 - ~/.nvm/versions/node/v19.8.1/bin/npm
    pnpm: 8.6.12 - ~/.nvm/versions/node/v19.8.1/bin/pnpm
    bun: 1.0.3 - ~/.bun/bin/bun
  Browsers:
    Chrome: 120.0.6099.109
    Edge: 120.0.2210.77
    Safari: 16.4
  npmPackages:
    typescript: ^5.3.3 => 5.3.3 
```

### Describe the bug

We startet migrating our nextjs project over to partially use the app directory. We use `"next": "^13.5.2"` in a turbo repo. The server/trpc setup is setup like it is described in the trpc docs. The only thing we did is to add a new util and a new provider for all the requests made from the app dir. Functionality wise this is working fine for client side requests.

Unfortunately the editors typescript language server is crashing all the time wich makes the dx horrible. It seems like type-checking is in a infinite loop because it takes a crazy amount of cpu.

Our code:

```ts
// src/app/_trpc/client.ts
import { createTRPCReact } from "@trpc/react-query";

import type { AppRouter } from "@acme/trpc";

export const trpc = createTRPCReact<AppRouter>({});
```

```tsx
// src/app/_trpc/provider.tsx
"use client";

import React, { useState } from "react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { httpBatchLink } from "@trpc/client";

import { transformer } from "@acme/trpc/transformer";

import { trpc } from "./client";

const getBaseUrl = () => {
  if (typeof window !== "undefined") return ""; // browser should use relative url
  if (process.env.VERCEL_URL) return `https://${process.env.VERCEL_URL}`; // SSR should use vercel url

  return `http://localhost:${process.env.PORT ?? 3000}`; // dev SSR should use localhost
};

export default function TrpcProvider({
  children,
}: {
  children: React.ReactNode;
}) {
  const [queryClient] = useState(() => new QueryClient());
  const [trpcClient] = useState(() =>
    trpc.createClient({
      transformer,
      links: [
        httpBatchLink({
          url: `${getBaseUrl()}/api/trpc`,
          fetch(url, options) {
            return fetch(url, {
              ...options,
              credentials: "include",
            });
          },
        }),
      ],
    }),
  );

  return (
    <trpc.Provider client={trpcClient} queryClient={queryClient}>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </trpc.Provider>
  );
}

```
```tsx
// src/app/layout.tsx
import TrpcProvider from "./_trpc/provider";


export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <TrpcProvider>
      <html lang="de" suppressHydrationWarning>
        <body>
          {children}
        </body>
      </html>
    </TrpcProvider>
  );
}
```

### Link to reproduction

https://stackblitz.com/edit/github-i7tdwc?file=src%2Fapp%2F_trpc%2Fclient.ts

### To reproduce

Make sure you use a proper amount of routes an procedures. We do have ~30 routes with 4 procedures.

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/5191">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/5191/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/trpc/trpc/issues/5191/pledge.svg">
</picture>
</a>

**Comment by KATT** [comment]
You could try using [`next`](https://github.com/trpc/trpc/tree/next) to see if it improves it. We've done a bunch of improvements in regards to this there & are planning to do even more.

Happy to chat more about your perf issues on Discord, feel free to DM me
